### PR TITLE
Add client diagnostics - Add sync APIs to RemoteFetcher

### PR DESCRIPTION
**Pull request by azabbasi**

### CODE REVIEW

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs [hunk 1]
```diff
@@ -16,20 +16,111 @@ namespace Azure.Iot.ModelsRepository.Fetchers
...
-            throw new NotImplementedException();
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope("RemoteModelFetcher.Fetch");
+            scope.Start();
+            try
+            {
+                Queue<string> work = PrepareWork(dtmi, repositoryUri);
+
+                string remoteFetchError = string.Empty;
+
+                while (work.Count != 0 && !cancellationToken.IsCancellationRequested)
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs:39]: If cancellation is requested, what is the right behavior from the called method? Short circuit as we've done here? Throw an [exception](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcanceledexception?view=net-5.0)?

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs [hunk 2]
```diff
@@ -16,20 +16,111 @@ namespace Azure.Iot.ModelsRepository.Fetchers
...
+                            CultureInfo.CurrentCulture,
+                            ServiceStrings.FetchingModelContent,
+                            tryContentPath));
+
+                    innerScope.Start();
+
+                    try
+                    {
+                        string content = EvaluatePath(tryContentPath, cancellationToken);
+                        return new FetchResult()
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs:54]: Parens not required

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs [hunk 3]
```diff
@@ -40,50 +131,88 @@ public async Task<FetchResult> FetchAsync(string dtmi, Uri repositoryUri, Cancel
...
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope("RemoteModelFetcher.EvaluatePath");
+            scope.Start();
+
+            try
+            {
+                using HttpMessage message = CreateGetRequest(path);
+
+                await _pipeline.SendAsync(message, cancellationToken).ConfigureAwait(false);
+
+                if (message.Response.Status >= 200 && message.Response.Status <= 299)
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs:181]: Probably should confirm there is a content stream. If the service returned 204, for instance, there wouldn't be a body.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs [hunk 4]
```diff
@@ -40,50 +131,88 @@ public async Task<FetchResult> FetchAsync(string dtmi, Uri repositoryUri, Cancel
...
+            request.Uri = uri;
 
-            Response response = await _pipeline.SendRequestAsync(request, cancellationToken).ConfigureAwait(false);
+            return message;
+        }
 
-            if (response.Status >= 200 && response.Status <= 299)
+        private static string GetContent(Stream content)
+        {
+            using (JsonDocument json = JsonDocument.Parse(content))
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs:211]: Want to set project lang to 8, so we can use the simplified `using` statement?

Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs:211]: We can, I changed it.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs (hunk 1 above)
Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs:39]: Good point, I think we should throw an exception with the correct message that cancellation was requested. changing.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs (hunk 3 above)
Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs:181]: Talked offline with Paymaun, he agrees that we should only accept 200's for the GET response. changing the code.

File: sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/LocalModelFetcher.cs [hunk 5]
```diff
@@ -27,35 +30,48 @@ public Task<FetchResult> FetchAsync(string dtmi, Uri repositoryUri, Cancellation
...
-            string fnfError = string.Empty;
-            while (work.Count != 0 && !cancellationToken.IsCancellationRequested)
-            {
-                string tryContentPath = work.Dequeue();
-                ResolverEventSource.Shared.FetchingModelContent(tryContentPath);
+                work.Enqueue(GetPath(dtmi, repositoryUri, false));
 
-                if (File.Exists(tryContentPath))
+                string fnfError = string.Empty;
+                while (work.Count != 0 && !cancellationToken.IsCancellationRequested)
```
Comment (drwill-ms) [inline sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/LocalModelFetcher.cs:48]: Should remove the check for IsCancellationRequested, no?

It'll throw on line 50 if it is, and that code won't get hit if we check for it here.

Comment (azabbasi) [reply sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/LocalModelFetcher.cs:48]: :D I forgot

**Review by drwill-ms — APPROVED** [review]

**Review by digimaun — APPROVED** [review]

**Pull request merged**